src/iter/parallel.rs

```diff
@@ -44,56 +43,61 @@ impl<'a, T: 'a + Send + Sync> UnindexedProducer for BitProducer<'a, T>
-                let usize_bits = size_of::<usize>() * 8;
-                let last_bit = usize_bits as u32 - self.0.masks[level].leading_zeros() - 1;
-                // If this is the highest level, there is no prefix saved as it's always zero
+            let mut handle_level = |level: usize| if self.0.masks[level] == 0 {
+                None
+            } else {
+                // Top levels prefix is zero because there is nothing before it
                 let level_prefix = self.0.prefix.get(level).cloned().unwrap_or(0);
-                // If there is one bit left, descend
-                if first_bit == last_bit {
-                    // Calculate the index based on prefix and the bit that is descended to
-                    let idx = (level_prefix | first_bit) as usize;
-                    // When descending all of the iteration happens in the child of the bit that is left
-                    self.0.prefix[level - 1] = (idx as u32) << BITS;
-                    // And because all the iteration happens in the child, it's parent can be removed.
-                    self.0.masks[level] = 0;
-                    // Get the mask of the child layer from the set
-                    self.0.masks[level - 1] = get_from_layer(self.0.set, level - 1, idx);
-                    return None;
-                }
-                // Make the split point to be the average of first and last bit.
-                let after_average = (first_bit + last_bit) / 2 + 1;
-                // A bit mask to get the lower half of the mask
-                // This cuts the mask to half from the average
-                let mask = (1 << after_average) - 1;
-                let mut other = BitProducer(BitIter::new(self.0.set, [0, 0, 0, 0], [0, 0, 0]));
-                let original_mask = self.0.masks[level];
-                // Take the higher half of the mask
-                other.0.masks[level] = original_mask & !mask;
-                // The higher levels of masks don't need to preserved as they are empty.
-                for n in &self.0.masks[(level + 1)..] {
-                    debug_assert_eq!(*n, 0);
-                }
-                // The higher half starts iterating after the average
-                other.0.prefix[level - 1] = (level_prefix | after_average) << BITS;
-                // And preserve the prefix of the higher levels
-                other.0.prefix[level..].copy_from_slice(&self.0.prefix[level..]);
-                // Take the lower half the mask
-                self.0.masks[level] = original_mask & mask;
-                // Combined mask of the current level should now equal the original mask
-                debug_assert_eq!(self.0.masks[level] | other.0.masks[level], original_mask);
-                // The lower half starts iterating from the first bit
-                self.0.prefix[level - 1] = (level_prefix | first_bit) << BITS;
-                Some(other)
+                let first_bit = self.0.masks[level].trailing_zeros();
```

Comment:
please note that for performance reasons (and this is a perf-critical code, if I understand correctly) we should be using unsafe getters in order to avoid bound checks for `[level]`

Reply:
A somewhat cleaner solution would be having each level with a struct (i.e. array of structs instead of structure of arrays `masks`, `prefix`, etc), and then pass references to the whole level around (previous and the current one).

Reply:
I don't think we need to use unsafe here, it's inside a closure and the index should really get inlined.

Reply:
As @torkleyy said this is called within closure and after inlining the `level` is a constant value.
Also `masks`, `prefix` are from `BitIter` so editing those would also change sequential iteration code.